docker/kuberay-autoscaler/autoscaling_config.py

```diff
@@ -0,0 +1,282 @@
+from contextlib import suppress
+import logging
+import math
+import os
+from typing import Dict, Optional
+
+import json
+
+from ray.autoscaler._private.providers import _get_node_provider
+from ray.autoscaler._private.util import validate_config
+
+logger = logging.getLogger(__name__)
+
+# The name of this Ray cluster. Should coincide with the `metadata.name` field of the
+# Ray CR.
+_RAY_CLUSTER_NAME = os.getenv("RAY_CLUSTER_NAME")
+
+# The Kubernetes namespace in which this Ray cluster runs.
+_RAY_CLUSTER_NAMESPACE = os.getenv("RAY_CLUSTER_NAMESPACE")
+
+# Logical group name for the KubeRay head group.
+# Used as the name of the "head node type" by the autoscaler.
+_HEAD_GROUP_NAME = "head-group"
+
+_GPU_WARNING_LOGGED = False
+
+
+def generate_autoscaling_config() -> dict:
+    """Generates an autoscaling config by reading the RayCluster CR and translating it
+    into a format readable by the autoscaler.
+
+    Used to fetch the autoscaling config at the beginning of each autoscaler iteration.
+
+    This function is the only public object in this file.
+    """
+    assert (
+        _RAY_CLUSTER_NAME
+    ), "The RAY_CLUSTER_NAME environment variable is unset. Check the RayCluster CR."
+    assert _RAY_CLUSTER_NAMESPACE, (
+        "The RAY_CLUSTER_NAMESPACE environment variable is unset. "
+        "Check the RayCluster CR."
+    )
+
+    provider_config = _generate_provider_config()
+
+    # Get the KubeRay node provider by instantiating it (on first call) or retrieving it
+    # from cache (subsequent calls).
+    # The autoscaler accesses the single node provider instance in the same way.
+    provider = _get_node_provider(
+        provider_config=provider_config, cluster_name=_RAY_CLUSTER_NAME, use_cache=True
+    )
+
+    # Fetch the Ray CR from K8s.
+    ray_cr = provider.get("rayclusters/{}".format(_RAY_CLUSTER_NAME))
```

Comment:
how does the provider detect the namespace?  

provider.get("rayclusters/{}/{}".format(_RAY_CLUSTER_NAMESPACE, _RAY_CLUSTER_NAME))

Reply:
I'm going to replace this part with a call to the Python K8s client.
Should be easier to read.

Reply:
Here is what I would do about this: Define a class
```
class AutoscalingConfigGenerator:
    def __init__(self):
        self.headers, self.verify = ray.autoscaler._private.kuberay.node_provider.load_k8s_secrets() # factor this method out

    def __call__(self, args):
        url = ray.autoscaler._private.kuberay.node_provider.url_from_resource("rayclusters/...") # factor out
        requests.get(url, headers=self.headers, verify=self.verify)
        # generate autoscaling config here
```
and then pass an instance of this class (which is a callable) to autoscaling_config of Monitor. This way, the secrets only have to be read once (vs. having to instantiate lots of providers or k8s clients, which is even worse than creating lots of providers) and there is maximum code sharing with the node provider/everything is consistent.

Reply:
Was planning to just instantiate a single extra global k8s client for this purpose, but sure.

Reply:
Created the class, named it `AutoscalingConfigProducer` to avoid suggesting that it's a Python generator. A few lines of code duplicated, but it's not so bad.